homeassistant/components/huawei_lte.py

```diff
@@ -0,0 +1,123 @@
+"""
+Support for Huawei LTE routers.
+
+For more details about this component, please refer to the documentation at
+https://home-assistant.io/components/huawei_lte/
+"""
+from datetime import timedelta
+from functools import reduce
+import logging
+import operator
+
+import voluptuous as vol
+import attr
+
+from homeassistant.const import (
+    CONF_URL, CONF_USERNAME, CONF_PASSWORD, EVENT_HOMEASSISTANT_STOP,
+)
+from homeassistant.helpers import config_validation as cv
+from homeassistant.util import Throttle
+
+
+_LOGGER = logging.getLogger(__name__)
+
+REQUIREMENTS = ['huawei-lte-api==1.0.12']
+
+MIN_TIME_BETWEEN_UPDATES = timedelta(seconds=10)
+
+DOMAIN = 'huawei_lte'
+DATA_KEY = 'huawei_lte'
+
+CONFIG_SCHEMA = vol.Schema({
+    DOMAIN: vol.All(cv.ensure_list, [vol.Schema({
+        vol.Required(CONF_URL): cv.url,
+        vol.Required(CONF_USERNAME): cv.string,
+        vol.Required(CONF_PASSWORD): cv.string,
+    })])
+}, extra=vol.ALLOW_EXTRA)
+
+
+@attr.s
+class RouterData:
+    """Class for router state."""
+
+    client = attr.ib()
+    device_information = attr.ib(init=False, factory=dict)
+    device_signal = attr.ib(init=False, factory=dict)
+    traffic_statistics = attr.ib(init=False, factory=dict)
+    wlan_host_list = attr.ib(init=False, factory=dict)
+
+    def __getitem__(self, path: str):
+        """
+        Get value corresponding to a dotted path.
+
+        The first path component designates a member of this class
+        such as device_information, device_signal etc, and the remaining
+        path points to a value in the member's data structure.
+        """
+        cat, *path_ = path.split(".")
+        return reduce(operator.getitem, path_, getattr(self, cat))
```

Comment:
This is very cryptic.

Reply:
I think that's somewhat of an exaggeration, but agree that how it works is likely not immediately clear to most readers on first casual read. Not sure what would be considerably cleaner though, suggestions? Perhaps something like this:

```python
        # ...
        def lookup(what, path):
            if not path:
                return what
            if not isinstance(what, dict):
                raise KeyError(path[0])
            return lookup(what[path[0]], path[1:])
        return lookup(getattr(self, cat), path_)
```

Reply:
Nah, I think the new version in the new PR is ok.